### PR TITLE
[inductor] Fix RNG order change caused by reorder_for_locality

### LINKED ISSUE #175156: [inductor] Multiple randint calls cause inconsistent RNG results between eager and compiled mode after fixing the rng seed
### 🐛 Describe the bug

To reproduce:
```python
import torch
import torch._dynamo
torch._inductor.config.fallback_random = True

def fn():
    torch.manual_seed(0)
    out = torch.randint(0, 100, (4,1), dtype=torch.int64)  # Shape and integer range are not important.
    other = torch.randint(0, 100, (2,1), dtype=torch.int64) # No inconsistency if remove this line, although the value won't be returned.
    return out
    
cfunc = torch.compile(fn, backend='inductor')
torch.manual_seed(0)
out1 = fn()
torch.manual_seed(0)
out2 = cfunc()

print(f'{out1=}')
'''
out1=tensor([[44],
        [39],
        [33],
        [60]])
'''
print(f'{out2=}')
'''
out2=tensor([[33],
        [60],
        [63],
        [79]])
'''

torch.testing.assert_close(out1, out2, equal_nan=True)
```

Problem:
RNG state is inconsistent between eager execution and torch.compile(backend='inductor') execution, even after explicitly resetting the seed with torch.manual_seed(0) before each run.
The issue occurs when two randint calls are called inside the compiled function:the output of the first randint already differs between eager and compiled mode, even though the second randint result is not returned.

After a few debugging, I notice that the following version does NOT trigger the inconsistency:

```python
def fn():
    torch.manual_seed(0)
    out = torch.randint(0, 100, (4,1), dtype=torch.int64)
    torch.manual_seed(0)  # Extra seed reset before second randint
    other = torch.randint(0, 100, (2,1), dtype=torch.int64)
    return out
```

The only difference is an extra torch.manual_seed(0) before the second randint.
This behavior is bit counterintuitive: We already set the seed before the first `randint`, and the second `randint` does not modify the returned value.
However, without setting the second `randint`'s rng seed still causes the first randint output to mismatch between eager and compiled mode.

Note that this inconsistency only occurs for `inductor` backend, but does not occur for `eager/aot_eager`backend.

### Error logs

_No response_

### Versions

[pip3] torch==2.11.0.dev20260209+cpu
[pip3] triton==3.6.0
[conda] torch 2.11.0.dev20260209+cpu pypi_0 pypi
[conda] triton 3.6.0 pypi_0 pypi

cc @ezyang @gchanan @kadeng @msaroufim @pbelevich @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by r83575**
Fixes #175156

## Problem

The `reorder_for_locality` pass in TorchInductor may reorder RNG operations
such as `aten.randint`.

Since RNG ops consume global RNG state, their execution order affects program
semantics. Reordering them can therefore produce results that differ from eager
execution.

As reported in the issue, two consecutive `torch.randint` calls may produce
different results between eager and compiled execution even when the RNG seed
is reset.

## Root cause

`reorder_for_locality` may move producer nodes using:

    node.prepend(other_node)

RNG operations are currently treated as pure ops and may therefore be reordered,
even though they consume RNG state.

## Fix

Prevent RNG ops from being reordered during the `reorder_for_locality` pass.

## Tests

Added a regression test:

    test/inductor/test_rng_reordering.py

which verifies that multiple `torch.randint` calls produce consistent results
between eager and compiled execution.

## Notes

This patch focuses on fixing the reported `aten.randint` case.
The same approach could potentially be generalized to other RNG ops
if necessary.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

### CODE REVIEW

File: test/inductor/test_rng_reordering.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Owner(s): ["module: inductor"]
+
+import torch
+from torch._inductor import config as inductor_config
+from torch._inductor.test_case import TestCase, run_tests
+
+
+class TestRNGReordering(TestCase):
```
Comment (Lucaskabela) [inline test/inductor/test_rng_reordering.py]: We can honestly fold this into an existing test file - no need to make a new file for a single test

File: torch/_inductor/fx_passes/post_grad.py
Comment (Lucaskabela) [inline torch/_inductor/fx_passes/post_grad.py]: This name should be more specific - we are only checking a single random op here right?

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Are we considering all random ops here? Or a large enough set of them?

**Comment by r83575** [comment]
@Lucaskabela Thanks for the review and the helpful suggestions.

The previous helper name was too broad, as it only matched `aten.randint.low`.

I updated the logic to use `torch.Tag.nondeterministic_seeded`, which is present
on RNG-state-consuming ops such as `aten.randint.low`, `aten.rand.default`, and
`aten.randn.default`. This makes the guard no longer limited to a single overload.

I also moved the regression test into `test/inductor/test_deterministic.py` and
kept the test name specific to `randint`, since that is the reported regression.

Please let me know if there’s anything else I should adjust.

File: test/inductor/test_rng_reordering.py (hunk 1 above)
Comment (r83575) [reply test/inductor/test_rng_reordering.py]: Thanks, agreed.

I moved the test into `test/inductor/test_deterministic.py` as suggested.

File: torch/_inductor/fx_passes/post_grad.py
Comment (r83575) [reply torch/_inductor/fx_passes/post_grad.py]: Good point.

I renamed the helper to `consumes_rng_state` to better reflect its semantics, and
updated the implementation to use `torch.Tag.nondeterministic_seeded` instead of
matching a single op. This makes it apply more generally to RNG-state-consuming ops,
not just `aten.randint.low`.

File: test/inductor/test_deterministic.py [hunk 2]
```diff
@@ -115,6 +115,25 @@ def foo(x):
             else:
                 self.assertTrue(counters["inductor"]["coordesc_tuning_bench"] > 0)
 
+    def test_reorder_for_locality_preserves_randint_order(self):
```
Comment (karthickai) [inline test/inductor/test_deterministic.py:118]: could you add @parametrize("fallback_random", [True, False]) to this test?

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -777,6 +777,13 @@ def check():
         def check():
             return True
 
+    def consumes_rng_state(node: torch.fx.Node) -> bool:
+        return (
+            node.op == "call_function"
+            and isinstance(node.target, torch._ops.OpOverload)
+            and torch.Tag.nondeterministic_seeded in getattr(node.target, "tags", ())
```
Comment (karthickai) [inline torch/_inductor/fx_passes/post_grad.py]: the line above already checks isinstance(node.target, torch._ops.OpOverload), and OpOverload alway has tags attribute (getattr fallback is unnecessary).

**Review by karthickai — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/post_grad.py (hunk 3 above)
Comment (r83575) [reply torch/_inductor/fx_passes/post_grad.py]: thanks!

I removed the `getattr` fallback and now check `node.target.tags` directly.

File: test/inductor/test_deterministic.py (hunk 2 above)
Comment (r83575) [reply test/inductor/test_deterministic.py:118]: Good suggestion, thanks.

Added parametrization over `fallback_random` to cover both modes.

File: test/inductor/test_deterministic.py [hunk 4]
```diff
@@ -115,6 +115,26 @@ def foo(x):
             else:
                 self.assertTrue(counters["inductor"]["coordesc_tuning_bench"] > 0)
 
+    @parametrize("fallback_random", [True, False])            
+    def test_reorder_for_locality_preserves_randint_order(self, fallback_random):
+        with inductor_config.patch(fallback_random=True):
```
Comment (karthickai) [inline test/inductor/test_deterministic.py:119]: update with `fallback_random` arg `inductor_config.patch(fallback_random=True)`

Comment (r83575) [reply test/inductor/test_deterministic.py:119]: Good catch, thanks. Updated to use the parametrized `fallback_random` value.

Comment (r83575) [reply test/inductor/test_deterministic.py:119]: I updated the test to use the parametrized `fallback_random` value and reran it.

With that change, the test passes for `fallback_random=True` but fails for `fallback_random=False`, so the current fix appears to address the reported fallback-random path only.

Do you think the `fallback_random=False` case should be handled in this PR as well, or would it be better to keep this PR scoped to the reported case and address that separately?

Comment (karthickai) [reply test/inductor/test_deterministic.py:119]: you can scope it only fallback_random=True, remove the parametrize over fallback_random.

Comment (r83575) [reply test/inductor/test_deterministic.py:119]: Sounds good, thanks. I’ll scope the test to `fallback_random=True` and remove the parametrization.

**Review by karthickai — APPROVED** [review]

**Review by Lucaskabela — APPROVED** [review]

**Review by jansel — APPROVED** [review]
